### DISCUSSION TITLE
[PyBullet]Try to implement spring constraint without damping

### DISCUSSION THREAD

**Discussion by legenary** [bulletphysics/bullet3#3740] [General]
Hello,

I am trying to implement a simple spring constraint between two multibody by creating a `JOINT_FIXED `constraint and update its `maxForce ` in every step simulation. I referred to #2100. However, everything works except there is damping observed in my final output. 

Let me elaborate. I put one cube at `x=0` and one cube at `x=5`. I then created a constraint in the middle:
`c_spring = p.createConstraint(box1ID, -1, box2ID, -1, p.JOINT_FIXED, [-1, 0, 0], [rest_length/2, 0, 0], [-rest_length/2, 0, 0])`
where the rest length of the spring was set to be 3.
(Btw, I'm having trouble understanding what `jointAxis` is for `JOINT_FIXED`, so I'm not sure whether I did the right thing here)

Then, before every step of the simulation, I updated the `maxForce` that this constraint can provide:
`maxForce = k * np.abs(np.linalg.norm(np.array(box1Position) - np.array(box2Position))-rest_length)` 
where `k` is the spring constant
Then, `p.changeConstraint(c_spring, maxForce=maxForce)`

The trajectories of the two cubes are:
![image](https://user-images.githubusercontent.com/38062004/92807876-ddc10780-f380-11ea-8f22-3242bc407197.png)

I expect no energy loss and perfect elastic deformation but instead some damping is observed. How do I get rid of this damping?

Thanks in advance!
Yifu

**Comment by merlinyx** [comment]
Hi, I'm also trying to implement some constraints between multibodies and thanks a lot for the post! (this is very late but I think the joint axis only matters for JOINT_GEAR and for other joint types using [0,0,0] is fine; in fact, using anything here is probably fine because I think it's ignored; and for the damping, maybe you'd do changeDynamics() for the multibodies? the default linear and angular damping coefficients are both 0.04. If you have figured out, I'd appreciate any updates here, thanks!)

**Reply by legenary** [reply]
Sorry I've already switched to C++ at this point; it allows more flexible implementations. I can't say for sure for PyBullet, but in Bullet C++ the damping coefficient for btRigidBody and btGeneric6DofSpringConstraint are separate things and they are both set to *no damping* by default. For btRigidBody it's 0, and for btGeneric6DofSpringConstraint it's 1.

**Reply by merlinyx** [reply]
I see, that makes sense! Thanks for the update!
